### PR TITLE
Add remote support

### PR DESCRIPTION
**Pull request by karol-bisztyga**
## Description

Add remote support

### CODE REVIEW

File: src/devTools.js [hunk 1]
```diff
@@ -25,12 +26,23 @@ function getRandomId() {
 }
 
 class DevToolsEnhancer {
-  constructor() {
-    this.enhance.updateStore = newStore => {
-      console.warn('devTools.updateStore is deprecated use composeWithDevTools instead: ' +
-        'https://github.com/zalmoxisus/remote-redux-devtools#use-devtools-compose-helper');
+    instance = {};
```
Comment (karol-bisztyga) [inline src/devTools.js]: invalid indent

File: src/devTools.js [hunk 2]
```diff
@@ -25,12 +26,23 @@ function getRandomId() {
...
+  errorCounts = {};
```
Comment (karol-bisztyga) [inline src/devTools.js:28]: this had to be added. The code before this change was incorrect. In a case when `socket.on('error')` was called before `socket.on('connect')`, `errorCounts` was undefined and there was an attempt of accessing its fields already, which resulted in a `TypeError`

File: src/devTools.js [hunk 3]
```diff
@@ -25,12 +26,23 @@ function getRandomId() {
...
+  urlPromise = null;
+  socketUrl = null;
```
Comment (karol-bisztyga) [inline src/devTools.js]: the result of the `urlPromise` used in `handleChange` - if it's not set, the function's execution's aborted.

File: src/devTools.js [hunk 4]
```diff
@@ -68,7 +80,7 @@ class DevToolsEnhancer {
       type,
       id: this.socket.id,
       name: this.instanceName,
-      instanceId: this.appInstanceId,
+      instanceId: this.instance.id,
```
Comment (karol-bisztyga) [inline src/devTools.js]: The difference between `this.instanceId` and `this.appInstanceId`:
- this.instanceId - can be specified in options.id, otherwise initialized with a socket.id or, if there's none, `getRandomId` on every `send()` call. `send` is used to send errors.
- this.appInstanceId - always initialized with `getRandomId`. Used in `relay` to emit messages to socket

So both of those fields are used to send data to the server(appInstanceId over the socket and instanceId over the HTTP(reporting an error)). The instance logs in to a server via sockets using appInstanceId and is then recognized by the server using that appInstanceId. I don't see why those two values should be separate and/or different.

File: src/devTools.js [hunk 5]
```diff
@@ -95,13 +107,31 @@ class DevToolsEnhancer {
...
+      store.liftedStore.dispatch({ type: 'IMPORT_STATE', ...nextLiftedState });
+      this.relay(
+        'STATE',
+        this.getLiftedState(store, instance.filters),
+        instance
+      );
+    } catch (e) {
+      this.relay('ERROR', e.message, instance);
+    }
+  };
```
Comment (karol-bisztyga) [inline src/devTools.js]: taken from [here](https://github.com/jhen0409/react-native-debugger/blob/639c02c0f3385c83e62b0ca483cb989db0e4b1a4/app/worker/reduxAPI.js#L133)

File: src/devTools.js [hunk 6]
```diff
@@ -95,13 +107,31 @@ class DevToolsEnhancer {
     }
   }
 
+  getLiftedState = (store, filters) => {
+    const fixedStore = store || this.store;
+    return filterStagedActions(fixedStore.liftedStore.getState(), filters);
+  };
```
Comment (karol-bisztyga) [inline src/devTools.js]: taken from [here](https://github.com/jhen0409/react-native-debugger/blob/639c02c0f3385c83e62b0ca483cb989db0e4b1a4/app/worker/reduxAPI.js#L69)

File: src/devTools.js [hunk 7]
```diff
@@ -95,13 +107,31 @@ class DevToolsEnhancer {
     }
   }
 
+  getLiftedState = (store, filters) => {
+    const fixedStore = store || this.store;
```
Comment (karol-bisztyga) [inline src/devTools.js]: this comes from the fact that I overwrote an existing `getLiftedState` function(the old one wouldn't receive any args). This code will be fixed.

File: src/devTools.js [hunk 8]
```diff
@@ -95,13 +107,31 @@ class DevToolsEnhancer {
...
+  };
+
   handleMessages = (message) => {
     if (
       message.type === 'IMPORT' || message.type === 'SYNC' && this.socket.id && message.id !== this.socket.id
     ) {
-      this.store.liftedStore.dispatch({
-        type: 'IMPORT_STATE', nextLiftedState: parse(message.state)
-      });
+      this.importPayloadFrom(this.store, message.state, this.instance);
```
Comment (karol-bisztyga) [inline src/devTools.js]: Apply a proper `import` functionality taken from `react-native-debugger`

File: src/devTools.js [hunk 9]
```diff
@@ -187,12 +217,36 @@ class DevToolsEnhancer {
...
+        .then((url) => {
+          this.socketOptions.hostname = url;
+          this.socketUrl = url;
+          this.start();
+        })
+        .catch((err) => {
+          console.log('Error obtaining socket url: ' + err.toString());
+        });
+    }
+  };
```
Comment (karol-bisztyga) [inline src/devTools.js]: this function has been added in order to resolve the promise and obtain a proper URL before connecting to the socket server

File: src/devTools.js [hunk 10]
```diff
@@ -187,12 +217,36 @@ class DevToolsEnhancer {
...
+
   start = () => {
     if (this.started || this.socket && this.socket.getState() === this.socket.CONNECTING) return;
 
     this.socket = socketCluster.connect(this.socketOptions);
 
-    this.socket.on('error', function (err) {
+    this.socket.on('error', (err) => {
```
Comment (karol-bisztyga) [inline src/devTools.js:244]: I think using `function` here causes an error when trying to refer to `this.errorCounts`. With lambda, we get the `this` from the outer scope.

File: src/devTools.js [hunk 11]
```diff
@@ -261,9 +315,10 @@ class DevToolsEnhancer {
   }
 
   enhance = (options = {}) => {
+    const instanceId = generateId(options.instanceId);
     this.init({
       ...options,
-      hostname: getHostForRN(options.hostname || 'localhost')
+      hostname: options.hostname,
```
Comment (karol-bisztyga) [inline src/devTools.js]: this is overwritten when resolving a promise anyway(the proper value would be in `socketOptions`)

File: src/devTools.js [hunk 12]
```diff
@@ -272,20 +327,25 @@ class DevToolsEnhancer {
...
+        this.store = configureStore(next, this.monitorReducer, {
+          maxAge,
+          trace: options.trace,
+          traceLimit: options.traceLimit,
+          shouldCatchErrors: !!this.sendOnError,
+          shouldHotReload: options.shouldHotReload,
+          shouldRecordChanges: options.shouldRecordChanges,
+          shouldStartLocked: options.shouldStartLocked,
+          pauseActionType: options.pauseActionType || '@@PAUSED',
+        })(reducer, initialState);
```
Comment (karol-bisztyga) [inline src/devTools.js]: this is just styling

File: src/devTools.js [hunk 13]
```diff
@@ -95,13 +110,26 @@ class DevToolsEnhancer {
...
+  };
+
   handleMessages = (message) => {
     if (
       message.type === 'IMPORT' || message.type === 'SYNC' && this.socket.id && message.id !== this.socket.id
     ) {
-      this.store.liftedStore.dispatch({
-        type: 'IMPORT_STATE', nextLiftedState: parse(message.state)
-      });
+      this.importPayloadFrom(this.store, message.state, this.instance);
```
Comment (Ashoat) [inline src/devTools.js]: 1. What's the point of supporting a custom `store` parameter in `getLiftedState` and `getLiftedStateRaw` if it's just set to `this.store` anyways?
2. What's the difference between `this.filters` and `this.instance.filters`? I can't find where `this.instance.filters` is set...

File: src/devTools.js [hunk 14]
```diff
@@ -130,8 +158,7 @@ class DevToolsEnhancer {
   };
 
   init(options) {
-    this.instanceName = options.name;
```
Comment (Ashoat) [inline src/devTools.js]: Why'd you get rid of the name assignment?

File: src/devTools.js (hunk 4 above)
Comment (Ashoat) [reply src/devTools.js]: I appreciate the critical thinking here. That said, I'd love if we could make this patch smaller so it will be easier to maintain. Would it have any negative effect to undo the following customizations you made:

1. Undoing the customization to use an object for `this.instance` instead of tracking things separately in `this.instanceName`, `this.instanceId`, etc.
2. Undoing the merging of `this.instanceId` and `this.appInstanceId`

File: src/devTools.js [hunk 15]
```diff
@@ -95,13 +110,26 @@ class DevToolsEnhancer {
     }
   }
 
+  importPayloadFrom = (store, state, instance) => {
+    try {
+      const nextLiftedState = importState(state, instance);
```
Comment (Ashoat) [inline src/devTools.js:111]: We could manually construct the `instance` here from `this.instanceId`, `this.instanceName`, etc.

File: src/devTools.js [hunk 16]
```diff
@@ -140,7 +167,7 @@ class DevToolsEnhancer {
     if (options.port) {
       this.socketOptions = {
         port: options.port,
-        hostname: options.hostname || 'localhost',
```
Comment (Ashoat) [inline src/devTools.js:143]: Why'd you remove the `|| 'localhost'` part?

File: src/remotedevDevtools.js [hunk 17]
```diff
@@ -0,0 +1,171 @@
+// code taken from
```
Comment (Ashoat) [inline src/remotedevDevtools.js]: 1. `remotedevDevtools` is a confusing name (doesn't mention Redux) and the pattern of capitalization is confusing.
2. It's not clear to me what this file is for... do we need all of it? And what for? Would it be better to just get it from the `remotedev` package? How much really needs to be customized?

File: src/devTools.js (hunk 12 above)
Comment (Ashoat) [reply src/devTools.js]: Please undo all styling changes to make the review easier

File: src/devTools.js [hunk 18]
```diff
@@ -25,24 +26,38 @@ function getRandomId() {
 }
 
 class DevToolsEnhancer {
-  constructor() {
-    this.enhance.updateStore = newStore => {
-      console.warn('devTools.updateStore is deprecated use composeWithDevTools instead: ' +
-        'https://github.com/zalmoxisus/remote-redux-devtools#use-devtools-compose-helper');
```
Comment (Ashoat) [inline src/devTools.js]: Please undo all styling changes to make the review easier

File: src/devTools.js (hunk 9 above)
Comment (Ashoat) [reply src/devTools.js]: For your new code, it would be great if you could maintain our coding conventions (eg. `return` on its own line within its own curly braces, avoid huge `else` block in favor of `return`ing in the preceding `if` block, etc.)

File: src/devTools.js (hunk 10 above)
Comment (Ashoat) [reply src/devTools.js:244]: Makes sense... surprised that mistake hasn't already been fixed!

File: src/devTools.js (hunk 11 above)
Comment (Ashoat) [reply src/devTools.js]: Makes sense... feel free to keep this change, despite my overall request for simplicity

File: src/devTools.js (hunk 2 above)
Comment (Ashoat) [reply src/devTools.js:28]: Good catch, surprised nobody noticed it before!

File: src/devTools.js (hunk 3 above)
Comment (Ashoat) [reply src/devTools.js]: How do we make sure that the connection happens after the URL is determined?

**Review by Ashoat — CHANGES_REQUESTED** [review]
I didn't get a chance to do a thorough review, but the most important feedback here is to please make the patch as small and as simple as possible. Start by removing all unnecessary/philosophical changes, and then try to refactor the remaining code with the goal of simplicity and minimizing the size of the patchfile.

File: src/devTools.js (hunk 3 above)
Comment (karol-bisztyga) [reply src/devTools.js]: `socketCluster.connect` is called in `this.start` which is called in `this.startWrapper`. In that last function we make sure we have a hostname(either it's provided or we use `Pomise.then`)

File: src/devTools.js (hunk 4 above)
Comment (karol-bisztyga) [reply src/devTools.js]: I see, I'll try to revert those things.

File: src/devTools.js (hunk 15 above)
Comment (karol-bisztyga) [reply src/devTools.js:111]: Yes. The reason behind removing `this.instanceId`, `this.instanceName` etc. in favor of `this.instance` was to use that object here. From one point of view I think this is a better approach but I understand your point that we want to do as small patch as possible, so reverting those change will force us to create an object here.

File: src/devTools.js (hunk 13 above)
Comment (karol-bisztyga) [reply src/devTools.js]: 1. right, this is redundant, I missed this.
2. it should be `this.filters`

File: src/devTools.js (hunk 14 above)
Comment (karol-bisztyga) [reply src/devTools.js]: in order to store it in `this.instance.name`. I moved this initialization [here](https://github.com/karol-bisztyga/remote-redux-devtools/pull/1/files#diff-c7675fb5800b150a641e9dcba518e608278e78d140ab84717aa61fbf5d88048eR339)

File: src/devTools.js (hunk 16 above)
Comment (karol-bisztyga) [reply src/devTools.js:143]: I removed it because if I didn't it wouldn't be possible to obtain it in `startWrapper` from the promise. We want to cover two situations:
- a developer passes a hostname when using this middleware - we want to use it right away for the connection
- a developer doesn't pass a hostname - we want to resolve the promise, obtain the hostname and overwrite `this.socketOptions.hostname`

In the first case the `options.hostname` would be `undefined`. We want to keep it that way to be able to recognize this scenario in `startWrapper`. If I didn't remove this here we'd never resolve the promise and go with `localhost` by default.

File: src/devTools.js (hunk 10 above)
Comment (karol-bisztyga) [reply src/devTools.js:244]: I just think for those two problems with `errorCounts` I've found: this code's author has probably never checked how the code behaves when the server's not up or there are some other problems...

File: src/devTools.js (hunk 11 above)
Comment (karol-bisztyga) [reply src/devTools.js]: I think a key here is a lack of `|| 'localhost'` as I explained in the other comment. I think this line is redundant however and can be removed - it does nothing really.

File: src/devTools.js (hunk 13 above)
Comment (karol-bisztyga) [reply src/devTools.js]: this also applies for `getLiftedState` and `getLiftedStateRaw`. In every case when we call them with args we pass values from `this` anyway.

File: src/devTools.js (hunk 4 above)
Comment (karol-bisztyga) [reply src/devTools.js]: applied according changes

File: src/devTools.js (hunk 9 above)
Comment (karol-bisztyga) [reply src/devTools.js]: ok, fixed

File: src/devTools.js (hunk 12 above)
Comment (karol-bisztyga) [reply src/devTools.js]: fixed

File: src/remotedevDevtools.js (hunk 17 above)
Comment (karol-bisztyga) [reply src/remotedevDevtools.js]: ad 1. I see, please note that `devTools.js` doesn't mention Redux either... I do not mind changing the name. Capitalization could be better(probably the uppercase `T`, like: `remotedevDevTools.js`). My thinking was: both files are originally called `devTools.js` so I wanted to put the name of the lib that it comes from as a prefix. If you have any better name for this file on your mind, please, let me know and I'm alright with changing it!

ad 2. Quick research on whether it is worth putting up remotedeDevtools.js in `redux-remote-devtools` patch:
- devTools.js - patches `__REDUX_DEVTOOLS_EXTENSION_COMPOSE__`
- remotedevDevtools.js - patches `__REDUX_DEVTOOLS_EXTENSION__`
- this file contains a little bit of a different logic than devTools.js yet it's not that different from what I saw. I think we should be able to replace `__REDUX_DEVTOOLS_EXTENSION__` with `__REDUX_DEVTOOLS_EXTENSION_COMPOSE__`, BUT we have to consider how `__REDUX_DEVTOOLS_EXTENSION__` is called:
    - for `@react-navigation/devtools` instance(`root.react.js`) - unfortunately here it is called inside of the code of navigation. So, to change that, we'd need to post a patch for the navigation devtools package.
    - for the `SquadCal` instance(`dev-tools.react.js`) - this one is done directly in our code, so that'd be easier.
- Having all that said, I think pasting the code of `remotedev` as it is now is a good solution - the API matches(with that subscribe logic that's there) and it's not a lot of code.
- I checked how it's done in `react-native-debugger` and it seems that they take the default export of `devTools.js` and assing the following functions to it: `send`, `connect`, `disconnect`(check [here](https://github.com/jhen0409/react-native-debugger/blob/639c02c0f3385c83e62b0ca483cb989db0e4b1a4/app/worker/index.js#L25)). I'm not sure why: only the `connect` function is used, so the way I implemented it(patching the runtime with an object containing only the `connect` field which is a function) omits unnecessary code and still works the same.
- BTW here is a PR with the changes that have been applied here: https://github.com/karol-bisztyga/remotedev/pull/1/files

File: src/devTools.js (hunk 16 above)
Comment (Ashoat) [reply src/devTools.js:143]: Makes sense, thanks for explaining!

File: src/devTools.js [hunk 19]
```diff
@@ -237,7 +287,7 @@ class DevToolsEnhancer {
   };
```
Comment (Ashoat) [inline src/devTools.js]: Ugh, GitHub doesn't let me comment on a line that's not in the PR. This comment was meant to be on line 232 (left) / 282 (right)... should that line reference to `this.start` be updated to `this.startWrapper`? We're not currently using `options.startOn`, but figured it wouldn't hurt to update it so it would work

File: src/devTools.js [hunk 20]
```diff
@@ -187,12 +211,38 @@ class DevToolsEnhancer {
...
+    // obtain the hostname
+    this.urlPromise
+      .then((url) => {
+        this.socketOptions.hostname = url;
+        this.socketUrl = url;
+        this.start();
+      })
+      .catch((err) => {
+        throw new Error('Error obtaining socket url: ' + err.toString());
+      });
```
Comment (Ashoat) [inline src/devTools.js]: Can we rewrite this with `async`/`await` syntax? Should be okay to make `startWrapper` an async function

File: src/devTools.js [hunk 21]
```diff
@@ -237,7 +287,7 @@ class DevToolsEnhancer {
   };
 
   handleChange(state, liftedState, maxAge) {
-    if (this.checkForReducerErrors(liftedState)) return;
+    if (this.checkForReducerErrors(liftedState) || !this.socketUrl) return;
```
Comment (Ashoat) [inline src/devTools.js]: 1. What's the issue with letting this code run before `this.socketUrl` is set?
2. I figured it would be more readable to check `this.started`, but I realized that doesn't get set until later. Wondering... does the issue (response to question 1) occur after `this.socketUrl` is set but before `this.started`? Or only before `this.socketUrl` is set?

File: src/remotedevDevtools.js (hunk 17 above)
Comment (Ashoat) [reply src/remotedevDevtools.js]: 1. Regarding the name, how about naming this file `connectToReduxDevServer.js`? Not sure, that might be too wordy...
2. Thanks for explaining this. I now understand why we have this and why it's necessary. I don't think we should replace `__REDUX_DEVTOOLS_EXTENSION__` with `__REDUX_DEVTOOLS_EXTENSION_COMPOSE__`
3. I think it would make more sense to include the `remotedev` package with a separate patchfile. That would we keep our patchfiles minimal, and makes it easier to get upstream changes that may come in future `remotedev` updates. (If we do this, ignore my comment above about the name)

File: src/remotedevDevtools.js [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+  } else socketOptions = defaultSocketOptions;
+  socket = socketCluster.create(socketOptions);
+  watch();
+}
+
+// obtains ulr from the promise and then performs the callback
+// also, overwrites `optionsObject.hostname`
+// obtained url is stored in `obtainedUrl` variable
+// if it's already obtained, skip fetching
+function obtainUrl(urlPromise, optionsObject, callback) {
```
Comment (Ashoat) [inline src/remotedevDevtools.js]: Instead of a function that takes a callback, I think this would make more sense as an async function. If we make `start` an `async` function, we can have it `await urlPromise`, and then `await handleURL(url, options)`. In this system we replace `obtainUrl` with `handleURL`... it's basically the same code, but I renamed it because it's no longer "obtaining" the URL by waiting on `urlPromise` (that would be handled by our new `async` `start` function)

**Review by Ashoat — CHANGES_REQUESTED** [review]
Thanks, the `devTools` file is much cleaner now! And thanks for your explanations... I think I understand how this code works a lot better now.

This is looking pretty good now. I have two high-level pieces of feedback:
1. Can we try to use `async`/`await` in this code to make it more readable?
2. I think it would make more sense to have two patchfiles, one on `remote-redux-devtools` and one on `remotedev`, and just install both packages in our `native/package.json`. That would allow us to keep our patches as minimal as possible, and make it easier to consume any upstream changes in `remotedev`.

I think once the stuff in this review is addressed we will basically be ready to ship this whole thing! (JSI + Redux debugging)

File: src/devTools.js (hunk 20 above)
Comment (karol-bisztyga) [reply src/devTools.js]: no problem

File: src/devTools.js (hunk 19 above)
Comment (karol-bisztyga) [reply src/devTools.js]: yes, right, I can change that

File: src/remotedevDevtools.js (hunk 17 above)
Comment (karol-bisztyga) [reply src/remotedevDevtools.js]: 1. You name it, I'm ok with this
2. cool
3. Got it, as you said, that discards the name case. I'll post a separate patch then.

File: src/devTools.js (hunk 21 above)
Comment (karol-bisztyga) [reply src/devTools.js]: I looked at this a bit and honestly, this seems to be some kind of leftover after my experiments. I checked and if we don't check for this(neither `started` nor `socketUrl`) it just proceeds to eventually call `this.socket.emit`. I think the socket will internally handle a case when there's no connection. I'll just remove that.

Edit: I checked and set `realtime: false` - this results in not setting a connection in `enhance` function. And what happened was the the instance was not visible in the monitor app but there were no errors, so I think it is as I expected - the socket tries to emit but internally handles the failure.

**Comment by karol-bisztyga** [comment]
> Thanks, the `devTools` file is much cleaner now! And thanks for your explanations... I think I understand how this code works a lot better now.
> 
> This is looking pretty good now. I have two high-level pieces of feedback:
> 
> 1. Can we try to use `async`/`await` in this code to make it more readable?
> 2. I think it would make more sense to have two patchfiles, one on `remote-redux-devtools` and one on `remotedev`, and just install both packages in our `native/package.json`. That would allow us to keep our patches as minimal as possible, and make it easier to consume any upstream changes in `remotedev`.
> 
> I think once the stuff in this review is addressed we will basically be ready to ship this whole thing! (JSI + Redux debugging)

Great! I applied proper changes!

File: src/devTools.js [hunk 23]
```diff
@@ -95,13 +101,31 @@ class DevToolsEnhancer {
...
 
+  importPayloadFrom = (state) => {
+    const instance = {
+      name: this.instanceName,
+      id: this.appInstanceId,
+      store: this.store,
+    };
+    try {
+      const nextLiftedState = importState(state, instance);
+      if (!nextLiftedState) return;
```
Comment (Ashoat) [inline src/devTools.js]: Nit: can we match the style in our repo? Ie. curly braces, and `return;` on its own line

File: src/devTools.js [hunk 24]
```diff
@@ -187,12 +211,36 @@ class DevToolsEnhancer {
...
+      throw new Error('no port provided');
+    }
+    if (this.socketOptions.hostname) {
+      // hostname provided - don't look for it
+      this.start();
+      return;
+    }
+    // obtain the hostname
+    try {
+      this.socketHostname = await this.hostnamePromise;
```
Comment (Ashoat) [inline src/devTools.js]: Why don’t we just assign to `this.socketOptions.hostname` directly?

**Review by Ashoat — APPROVED** [review]
This looks great! If you can just address my two nits below, merge the PR, and then update the patchfile on Phabricator, we should be good to go :)

PS – can you mention the build command you’ve been using to convert from `src` to `lib`?

**Comment by karol-bisztyga** [comment]
> This looks great! If you can just address my two nits below, merge the PR, and then update the patchfile on Phabricator, we should be good to go :)
> 
> PS – can you mention the build command you’ve been using to convert from `src` to `lib`?

Sure!
The command is `yarn build` from the project's root dir. The build is also run after installing dependencies with `yarn`.

File: src/devTools.js (hunk 23 above)
Comment (karol-bisztyga) [reply src/devTools.js]: sure

File: src/devTools.js (hunk 24 above)
Comment (karol-bisztyga) [reply src/devTools.js]: right, seems like `socketHostname` is no longer needed and can be removed, thanks.
